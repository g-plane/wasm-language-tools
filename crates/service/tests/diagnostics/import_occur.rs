use super::*;
use insta::assert_json_snapshot;
use lsp_types::Uri;
use wat_service::LanguageService;

#[test]
fn empty() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "(module)";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}

#[test]
fn no_imports() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "(module (func))";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}

#[test]
fn single() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = r#"
(module
  (import "a" "b" (func))
  (func))
"#;
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}

#[test]
fn multi() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = r#"
(module
  (import "a" "b" (func))
  (import "c" "d" (func))
  (func))
"#;
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}

#[test]
fn single_after_other_fields() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = r#"
(module
  (func)
  (import "a" "b" (func)))
"#;
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn multi_after_other_fields() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = r#"
(module
  (func)
  (import "a" "b" (func))
  (import "c" "d" (func)))
"#;
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}
