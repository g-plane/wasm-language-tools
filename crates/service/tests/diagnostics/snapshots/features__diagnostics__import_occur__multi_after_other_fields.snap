---
source: crates/service/tests/diagnostics/import_occur.rs
expression: response
---
{
  "kind": "full",
  "items": [
    {
      "range": {
        "start": {
          "line": 3,
          "character": 2
        },
        "end": {
          "line": 3,
          "character": 25
        }
      },
      "severity": 1,
      "code": "import-occurrence",
      "source": "wat",
      "message": "import must occur before all non-import definitions"
    }
  ]
}
